src/ofonotracker.cpp

```diff
@@ -0,0 +1,335 @@
+    m_systemBus.connect(OFONO_SERVICE, OFONO_MANAGER_OBJECT, OFONO_MANAGER_INTERFACE, OFONO_MANAGER_MODEM_REMOVED,
+                        this, SLOT(handleManagerModemRemoved(QDBusObjectPath)));
+    m_systemBus.connect(OFONO_SERVICE, QString(), OFONO_MODEM_INTERFACE, OFONO_MODEM_PROPERTY_CHANGED,
+                        this, SLOT(handleModemPropertyChanged(QString, QDBusVariant)));
+    /* For the benefit of existing code that assumes also D-Bus based values
+     * to be available immediately after the object is instantiated -> make
+     * it possible to utilize a synchronous probe and evaluate property values
+     * before returning from constructor. */
+    if (synchronousInit)
+        getModemsSync();
+    else
+        getModemsAsync();
+}
+
+OFonoTrackerPrivate::~OFonoTrackerPrivate()
+{
+}
+
+QStringList OFonoTrackerPrivate::imeiNumbers() const
+{
+    qWarning() << "#### call imeiNumbers() ->" << m_imeiNumbers;
+    return m_imeiNumbers;
+}
+
+void OFonoTrackerPrivate::evaluateProperties()
+{
+    qWarning() << "#### evaluate modem properties";
+    QStringList imeiNumbers;
+    for (auto iter = m_modemPathList.cbegin(); iter != m_modemPathList.cend(); ++iter) {
+        QString imei(m_modemInfoHash[*iter].m_properties.value(OFOFO_MODEM_PROPERTY_SERIAL).toString());
+        if (!imei.isEmpty())
+            imeiNumbers << imei;
+    }
+    if (m_imeiNumbers != imeiNumbers) {
+        m_imeiNumbers = imeiNumbers;
+        qWarning() << "#### emit imeiNumbersChanged() ->" << m_imeiNumbers;
+        Q_Q(OFonoTracker);
+        Q_EMIT q->imeiNumbersChanged();
+    }
+    m_evaluatePropertiesTimer->stop();
+}
+
+void OFonoTrackerPrivate::handleManagerModemAdded(QDBusObjectPath objectPath, QVariantMap properties)
+{
+    addModem(ModemInfo(objectPath.path(), properties));
+}
+
+void OFonoTrackerPrivate::handleManagerModemRemoved(QDBusObjectPath objectPath)
+{
+    removeModem(objectPath.path());
+}
+
+void OFonoTrackerPrivate::handleModemPropertyChanged(QString key, QDBusVariant value)
+{
+    if (calledFromDBus())
+        setModemProperty(message().path(), key, value.variant());
+}
+
+void OFonoTrackerPrivate::evaluatePropertiesLater()
+{
+    if (!m_evaluatePropertiesTimer->isActive())
+        m_evaluatePropertiesTimer->start();
+}
+
+void OFonoTrackerPrivate::getModemsAsync()
+{
+    QDBusMessage request(QDBusMessage::createMethodCall(OFONO_SERVICE, OFONO_MANAGER_OBJECT, OFONO_MANAGER_INTERFACE, OFONO_MANAGER_GET_MODEMS));
+    request.setAutoStartService(false);
+    QDBusPendingReply<ModemInfoList> call(m_systemBus.asyncCall(request));
+    QDBusPendingCallWatcher *watcher = new QDBusPendingCallWatcher(call, this);
+    connect(watcher, &QDBusPendingCallWatcher::finished, this, [this](QDBusPendingCallWatcher *watcher) {
+        QDBusPendingReply<ModemInfoList> response(*watcher);
+        if (response.isValid())
+            setModems(response.value());
+        else if (response.error().name() != DBUS_ERROR_NAME_HAS_NO_OWNER)
```

Comment:
Just a comment: I find it strange that this error doesn't have a value in [ErrorType](https://doc.qt.io/archives/qt-5.6/qdbuserror.html#ErrorType-enum).